nni/algorithms/compression/v2/pytorch/base/pruner_tools.py

```diff
@@ -0,0 +1,360 @@
+            The compressor binded with this DataCollector.
+        """
+        self.compressor = compressor
+
+    def reset(self):
+        """
+        Reset the `DataCollector`.
+        """
+        raise NotImplementedError()
+
+    def collect(self) -> Dict:
+        """
+        Collect the compressor needed data, i.e., module weight, the output of activation function.
+
+        Returns
+        -------
+        Dict
+            Usually has format like {module_name: tensor_type_data}.
+        """
+        raise NotImplementedError()
+
+
+class HookCollectorInfo:
+    def __init__(self, targets: Union[Dict[str, Tensor], List[LayerInfo]], hook_type: str,
+                 collector: Union[Callable[[List, Tensor], Callable[[Tensor], None]], Callable[[List], Callable[[Module, Tensor, Tensor], None]]]):
+        """
+        This class used to aggregate the information of what kind of hook is placed on which layers.
+
+        Parameters
+        ----------
+        targets
+            List of LayerInfo or Dict of {layer_name: weight_tensor}, the hook targets.
+        hook_type
+            'forward' or 'backward'.
+        collector
+            A hook function generator, the input is a buffer (empty list) or a buffer (empty list) and tensor, the output is a hook function.
+            The buffer is used to store the data wanted to hook.
+        """
+        self.targets = targets
+        self.hook_type = hook_type
+        self.collector = collector
+
+
+class TrainerBasedDataCollector(DataCollector):
+    """
+    This class includes some trainer based util functions, i.e., patch optimizer or criterion, add hooks.
+    """
+
+    def __init__(self, compressor: Compressor, trainer: Callable[[Module, Optimizer, Callable], None], optimizer: Optimizer,
+                 criterion: Callable[[Tensor, Tensor], Tensor], training_epochs: int,
+                 opt_before_tasks: List = [], opt_after_tasks: List = [],
+                 collector_infos: List[HookCollectorInfo] = [], criterion_patch: Callable[[Callable], Callable] = None):
+        """
+        Parameters
+        ----------
+        compressor
+            The compressor binded with this DataCollector.
+        trainer
+            A callable function used to train model or just inference. Take model, optimizer, criterion as input.
+
+            Example::
+
+                def trainer(model: Module, optimizer: Optimizer, criterion: Callable[[Tensor, Tensor], Tensor]):
+                    model.train()
+                    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+                    for batch_idx, (data, target) in enumerate(train_loader):
+                        data, target = data.to(device), target.to(device)
+                        optimizer.zero_grad()
+                        output = model(data)
+                        loss = criterion(output, target)
+                        loss.backward()
+                        optimizer.step()
```

Comment:
Should we recommend the user to write the trainer so that parameter update is performed during data collection?

Reply:
It is a good suggestion, add it

Reply:
I mean do we want the user to always call optimizer.step() even when this trainer is used for calculating scores?

Reply:
got it, add a comment for this, this trainer is always used for data collection.